### PR TITLE
GraphAuthenticationProvider checks the RequestInformation.URI before appending the header

### LINKED ISSUE #2710: GraphAuthenticationProvider should check the requestinformation URI before appending the header
### Microsoft.Identity.Web Library

Microsoft.Identity.Web.Graph

### Microsoft.Identity.Web version

2.17.0

### Web app

Not Applicable

### Web API

Not Applicable

### Token cache serialization

Not Applicable

### Description

GraphAuthenticationProvider from Microsoft.Identity.Web.GraphServiceClient should validate that the request uri is a valid graph URI before setting the Authorization header. Otherwise, requests made with the graph client in scenarios such as large file upload will cause undesired behaviour. See https://github.com/microsoftgraph/msgraph-sdk-dotnet/issues/2379

This can also be done by using the [AllowedHostValidator](https://github.com/microsoft/kiota-abstractions-dotnet/blob/00a8f957808246c3b2bebba474655dd8b16a5d4d/src/authentication/AllowedHostsValidator.cs#L51) using the config at https://github.com/microsoftgraph/msgraph-sdk-dotnet-core/blob/5915858fccbb0ff0f7829c27c722bd2dcabc2e48/src/Microsoft.Graph.Core/Authentication/AzureIdentityAccessTokenProvider.cs#L19

### Reproduction steps

See code sample at https://github.com/microsoftgraph/msgraph-sdk-dotnet/issues/2379

### Error message

```shell
Unhandled exception. Microsoft.Identity.Web.MicrosoftIdentityWebChallengeUserException: IDW10502: 
An MsalUiRequiredException was thrown due to a challenge for the user.
See https://aka.ms/ms-id-web/ca_incremental-consent.
```

### Id Web logs

_No response_

### Relevant code snippets

```csharp
static async Task UploadDriveItem(GraphServiceClient graphServiceClient, Drive drive, string filePath, string itemPath)
{
	Console.WriteLine($"Uploading {filePath} to {itemPath}");
	using var fileStream = File.OpenRead(filePath);


	var uploadSessionRequestBody = new DriveUpload.CreateUploadSessionPostRequestBody
	{
		Item = new DriveItemUploadableProperties
		{
			AdditionalData = new System.Collections.Generic.Dictionary<string, object>
			{
				{ "@microsoft.graph.conflictBehavior", "replace" },
			},
		},
	};

	// Create the upload session
	// itemPath does not need to be a path to an existing item
	// var myDrive = await graphServiceClient.Me.Drive.GetAsync(r => r.Options.WithAppOnly());
	var uploadSession = await graphServiceClient.Drives[drive.Id]
		.Items["root"]
		.ItemWithPath(itemPath)
		.CreateUploadSession
		.PostAsync(uploadSessionRequestBody, r => r.Options.WithAppOnly());

	// Max slice size must be a multiple of 320 KiB
	int maxSliceSize = 320 * 1024;
	var fileUploadTask = new LargeFileUploadTask<DriveItem>(
		uploadSession, fileStream, maxSliceSize, graphServiceClient.RequestAdapter);

	var totalLength = fileStream.Length;
	// Create a callback that is invoked after each slice is uploaded
	IProgress<long> progress = new Progress<long>(prog =>
	{
		Console.WriteLine($"Uploaded {prog} bytes of {totalLength} bytes");
	});

	try
	{
		// Upload the file
		var uploadResult = await fileUploadTask.UploadAsync(progress);

		Console.WriteLine(uploadResult.UploadSucceeded ?
			$"Upload complete, item ID: {uploadResult.ItemResponse.Id}" :
			"Upload failed");
	}
	catch (Microsoft.Graph.Models.ODataErrors.ODataError ex)
	{
		Console.WriteLine($"Error uploading: {ex.Error?.Message}");
	}
}
```

### Regression

_No response_

### Expected behavior

Requests made to non-graph urls should not attempt to acquire the token.

### PR DESCRIPTION
**Pull request by kellyyangsong**
Fixes #2710

### CODE REVIEW

File: tests/E2E Tests/GraphServiceClientTests/GraphServiceClientTests.cs [hunk 1]
```diff
@@ -67,5 +72,45 @@ async Task Test()
...
+        [Fact]
+        public async Task AuthenticateRequestAsync_GraphUri_SetsAuthZHeader()
+        {
+            // arrange
+            RequestInformation request = new()
+            {
+                URI = new Uri("https://graph.microsoft.com/.default")
+            };
+
+            var mockAuthorizationHeaderProvider = new Mock<IAuthorizationHeaderProvider>();
```
Comment (kellyyangsong) [inline tests/E2E Tests/GraphServiceClientTests/GraphServiceClientTests.cs:103]: Tried to setup an IAuthorizationHeaderProvider like so: 

```csharp
TokenAcquirerFactory tokenAcquirerFactory = TokenAcquirerFactory.GetDefaultInstance();
IServiceCollection services = tokenAcquirerFactory.Services;

services.Configure<MicrosoftIdentityApplicationOptions>(s_optionName, option =>
    {
        option.Instance = "https://login.microsoftonline.com/";
        option.TenantId = "msidlab4.onmicrosoft.com";
        option.ClientId = "f6b698c0-140c-448f-8155-4aa9bf77ceba";
    });
IServiceProvider serviceProvider = tokenAcquirerFactory.Build();
IAuthorizationHeaderProvider authorizationHeaderProvider = serviceProvider.GetRequiredService<IAuthorizationHeaderProvider>();         
```
However, couldn't seem to get it to work. Set up a mock for now to show that the code works as intended and that an authorization header is appropriately added when RequestInformation.URI is a graph URI.

File: tests/E2E Tests/GraphServiceClientTests/GraphServiceClientTests.csproj [hunk 2]
```diff
@@ -7,6 +7,7 @@
 
   <ItemGroup>
     <PackageReference Include="Microsoft.NET.Test.Sdk" Version="$(MicrosoftNetTestSdkVersion)" />
+    <PackageReference Include="Moq" Version="4.20.70" />
```
Comment (kellyyangsong) [inline tests/E2E Tests/GraphServiceClientTests/GraphServiceClientTests.csproj:10]: Will remove Moq or pull out version number into Directory.Build.Props

File: tests/E2E Tests/GraphServiceClientTests/GraphServiceClientTests.cs (hunk 1 above)
Comment (kellyyangsong) [reply tests/E2E Tests/GraphServiceClientTests/GraphServiceClientTests.cs:103]: Discussed offline with @jmprieur - removing as we have plenty of positive tests for the graph service client

File: tests/E2E Tests/GraphServiceClientTests/GraphServiceClientTests.csproj (hunk 2 above)
Comment (kellyyangsong) [reply tests/E2E Tests/GraphServiceClientTests/GraphServiceClientTests.csproj:10]: Removing moq

File: src/Microsoft.Identity.Web.GraphServiceClient/GraphAuthenticationProvider.cs [hunk 3]
```diff
@@ -79,8 +79,11 @@ internal class GraphAuthenticationProvider : IAuthenticationProvider
                 authorizationHeaderProviderOptions = graphServiceClientOptions;
             }
 
+            AllowedHostsValidator allowedHostsValidator = new(
+                ["graph.microsoft.com", "graph.microsoft.us", "dod-graph.microsoft.us", "graph.microsoft.de", "microsoftgraph.chinacloudapi.cn", "canary.graph.microsoft.com", "graph.microsoft-ppe.com"]);
```
[the remark this answers is not in the document]
Comment (kellyyangsong) [reply src/Microsoft.Identity.Web.GraphServiceClient/GraphAuthenticationProvider.cs:83]: Done

File: src/Microsoft.Identity.Web.GraphServiceClient/GraphAuthenticationProvider.cs [hunk 4]
```diff
@@ -22,6 +22,7 @@ internal class GraphAuthenticationProvider : IAuthenticationProvider
         private const string AuthorizationHeaderKey = "Authorization";
         readonly IAuthorizationHeaderProvider _authorizationHeaderProvider;
         readonly GraphServiceClientOptions _defaultAuthenticationOptions;
+        private readonly string[] _graphUris = ["graph.microsoft.com", "graph.microsoft.us", "dod-graph.microsoft.us", "graph.microsoft.de", "microsoftgraph.chinacloudapi.cn", "canary.graph.microsoft.com", "graph.microsoft-ppe.com"];
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web.GraphServiceClient/GraphAuthenticationProvider.cs:25]: how do we know these are all the URIs?

File: src/Microsoft.Identity.Web.GraphServiceClient/GraphAuthenticationProvider.cs [hunk 5]
```diff
@@ -79,8 +80,9 @@ internal class GraphAuthenticationProvider : IAuthenticationProvider
                 authorizationHeaderProviderOptions = graphServiceClientOptions;
             }
 
+            AllowedHostsValidator allowedHostsValidator = new(_graphUris);
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web.GraphServiceClient/GraphAuthenticationProvider.cs:83]: this is cool. had to read up on what this is.

**Review by jennyf19 — APPROVED** [review]
:shipit:

File: src/Microsoft.Identity.Web.GraphServiceClient/GraphAuthenticationProvider.cs (hunk 4 above)
Comment (kellyyangsong) [reply src/Microsoft.Identity.Web.GraphServiceClient/GraphAuthenticationProvider.cs:25]: @jennyf19 great question - check out [AzureIdentityAccessTokenProvider.cs](https://github.com/microsoftgraph/msgraph-sdk-dotnet-core/blob/5915858fccbb0ff0f7829c27c722bd2dcabc2e48/src/Microsoft.Graph.Core/Authentication/AzureIdentityAccessTokenProvider.cs#L19) from the Graph repo, shared by a friend from Graph @andrueastman

**Review by andrueastman — APPROVED** [review]
Thanks for making this change!

**Review by keegan-caruso — APPROVED** [review]

**Review by pmaytak — APPROVED** [review]

**Pull request merged**